theano/compile/function_module.py

```diff
@@ -853,14 +859,17 @@ def __call__(self, *args, **kwargs):
         if not self.trust_input:
             for c in self.input_storage:
                 if c.required and not c.provided:
+                    restore_defaults()
```

Comment:
I think line 813 need a call like this before the raise. We catch and extend an exception.
I think line 898 also need a call to this. But for that one, it would be better to add the call just after the "except" as the other branch will raise the error itself and won't go up to line 898.

Can you finish that or we let this for another PR?